Userland/Utilities/tsort.cpp

```diff
@@ -0,0 +1,145 @@
+
+        // If a node has already been prioritized, it can be ignored.
+        if (node.status == NodeStatus::Prioritized) {
+            stack.take_last();
+            continue;
+        }
+
+        // Keep track of which nodes have been seen to detect cycles.
+        node.status = NodeStatus::Seen;
+
+        if (node.ancestors.is_empty()) {
+            // If a node has no remaining ancestors (dependencies), it either never had
+            // ancestors, or its ancestors have already been prioritized. In either case,
+            // this is now the deepest un-prioritized node, which makes it the next
+            // highest priority.
+            node.status = NodeStatus::Prioritized;
+            outln("{}", stack.take_last().name);
+        } else {
+            auto next_ancestor_name = node.ancestors.pop();
+            auto next_ancestor = node_map.get(next_ancestor_name).release_value();
+            if (next_ancestor.status == NodeStatus::Seen)
+                // If the same node is seen multiple times, this represents a cycle in
+                // the graph. To avoid an infinite loop, the duplicate node is not added
+                // to the stack a second time. Instead, the edge is deliberately ignored,
+                // and the topological sort proceeds as though the cycle did not exist.
+
+                handle_cycle(stack, next_ancestor);
+            else
+                // Recursively prioritize all ancestors.
+                stack.append(next_ancestor);
+        }
+    }
+}
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    TRY(Core::System::pledge("stdio rpath"));
+
+    StringView path;
+    bool quiet = false;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(path, "Path to file", "path", Core::ArgsParser::Required::No);
+    args_parser.add_option(quiet, "Suppress warnings about cycles", "quiet", 'q');
+    args_parser.parse(arguments);
+
+    auto file = TRY(Core::Stream::File::open_file_or_standard_stream(path, Core::Stream::OpenMode::Read));
+    auto input_bytes = TRY(file->read_all());
+    auto inputs = StringView(input_bytes).split_view_if(is_ascii_space);
+
+    if (inputs.is_empty())
+        return 0;
+
+    if (inputs.size() % 2 != 0) {
+        warnln("tsort: the number of inputs must be even");
+        return 1;
+    }
+
+    NodeMap node_map;
+
+    // Each pair of inputs (e.g. "a b") represents an edge of a directed acyclic graph.
+    // If the same input is repeated (e.g. "a a"), this defines a single node with no
+    // connection to any other nodes. Otherwise, the first input is interpreted as an
+    // ancestor of the second.
+    for (size_t i = 0; i < inputs.size(); i += 2) {
+        auto ancestor_name = inputs[i];
+        auto descendent_name = inputs[i + 1];
+
+        auto& descendent_node = node_map.ensure(descendent_name, [&]() { return Node { descendent_name, HashTable<StringView> {}, NodeStatus::NotSeen }; });
```

Comment:
This is a bit hidden by the API, but `HashMap::ensure()` calls through to `HashTable::set()`, which does `MUST(try_set(...))`, so this hides a possible allocation error.

Optional to fix this since this is an issue elsewhere I'm sure, so feel free to resolve this as-is, but good to be aware that this might crash in OOM situations. Ideally we would have a `HashMap::try_ensure()` that uses `try_set()` and can return an error. A little yakbait if you want to take it I suppose :^)